src/middlewared/middlewared/plugins/crypto.py

```diff
@@ -101,60 +237,862 @@ def cert_issuer(cert):
-                cert['cert_CSR'] = crypto.dump_certificate_request(crypto.FILETYPE_PEM, csr_obj).decode()
-        except:
-            self.logger.debug('Failed to load csr {0}'.format(cert['cert_name']), exc_info=True)
+            if cert['CSR']:
+                csr_obj = crypto.load_certificate_request(crypto.FILETYPE_PEM, cert['CSR'])
+                cert['CSR'] = crypto.dump_certificate_request(crypto.FILETYPE_PEM, csr_obj).decode()
+        except Exception:
+            self.logger.debug('Failed to load csr {0}'.format(cert['name']), exc_info=True)
 
-        cert['cert_internal'] = 'NO' if cert['cert_type'] in (CA_TYPE_EXISTING, CERT_TYPE_EXISTING) else 'YES'
+        cert['internal'] = 'NO' if cert['type'] in (CA_TYPE_EXISTING, CERT_TYPE_EXISTING) else 'YES'
 
         obj = None
         # date not applicable for CSR
-        cert['cert_from'] = None
-        cert['cert_until'] = None
-        if cert['cert_type'] == CERT_TYPE_CSR:
+        cert['from'] = None
+        cert['until'] = None
+        if cert['type'] == CERT_TYPE_CSR:
             obj = csr_obj
         elif cert_obj:
             obj = cert_obj
             notBefore = obj.get_notBefore()
             t1 = dateutil.parser.parse(notBefore)
             t2 = t1.astimezone(dateutil.tz.tzutc())
-            cert['cert_from'] = t2.ctime()
+            cert['from'] = t2.ctime()
 
             notAfter = obj.get_notAfter()
             t1 = dateutil.parser.parse(notAfter)
             t2 = t1.astimezone(dateutil.tz.tzutc())
-            cert['cert_until'] = t2.ctime()
+            cert['until'] = t2.ctime()
 
         if obj:
-            cert['cert_DN'] = '/' + '/'.join([
+            cert['DN'] = '/' + '/'.join([
                 '%s=%s' % (c[0].decode(), c[1].decode())
                 for c in obj.get_subject().get_components()
             ])
 
         return cert
 
+    # HELPER METHODS
+
+    @accepts(
+        Str('certificate', required=True)
+    )
+    async def load_certificate(self, certificate):
+        try:
+            cert = crypto.load_certificate(
+                crypto.FILETYPE_PEM,
+                certificate
+            )
+        except crypto.Error:
+            return {}
+        else:
+            cert_info = {
+                'country': cert.get_subject().C,
+                'state': cert.get_subject().ST,
+                'city': cert.get_subject().L,
+                'organization': cert.get_subject().O,
+                'common': cert.get_subject().CN,
+                'san': cert.get_subject().subjectAltName,
+                'email': cert.get_subject().emailAddress,
+            }
+
+            signature_algorithm = cert.get_signature_algorithm().decode()
+            m = re.match('^(.+)[Ww]ith', signature_algorithm)
+            if m:
+                cert_info['digest_algorithm'] = m.group(1).upper()
+
+            return cert_info
+
+    @accepts(
+        Str('buffer', required=True),
+        Str('passphrase', default=None)
+    )
```

Comment:
Why does this need to be a public exposed method?

Reply:
Do you think this could be used outside this service ? We are using this in Certificate Authority Service though. But if it is only going to have two usages, we can move it outside the class like the common validate method we have

Reply:
Good question, do you think ther ewould be any use for the end user to dump the private key of any certificate?

Reply:
Personally, it seems unlikely that an end user will want to use this. The only possibility that this may be used somewhere is if we use it in any other services we have.

Reply:
Then lets just mark it with @private so its not shown in the API docs.